### PR TITLE
Remove grunt

### PR DESCRIPTION
**Pull request by noppanit**
As we don't need Grunt and Gulp, this PR attempts to replace Grunt with Gulp.

Related: https://github.com/HabitRPG/habitrpg/issues/5829

### CODE REVIEW

File: tasks/gulp-build.js [hunk 1]
```diff
@@ -1,4 +1,12 @@
 import gulp from 'gulp';
+import browserify from 'browserify';
+
+let source = require('vinyl-source-stream');
+let rename = require("gulp-rename");
+let minifyCss = require('gulp-minify-css');
+let stylus = require('gulp-stylus');
+let nib = require('nib');
+
 require('gulp-grunt')(gulp);
```
Comment (crookedneighbor) [inline tasks/gulp-build.js]: Is this needed?

File: tasks/gulp-build.js [hunk 2]
```diff
@@ -10,7 +18,36 @@ gulp.task('build', () => {
...
 
 gulp.task('build:dev', ['babel:common', 'prepare:staticNewStuff'], (done) => {
-  gulp.start('grunt-build:dev', done);
+	let b = browserify({
+		entries: ['./common/index.js'],
+	});
+
+	b.transform('coffeeify').bundle().pipe(source('index.js')).pipe(rename('habitrpg-shared.js'))
+		.pipe(gulp.dest('common/dist/scripts/'))
+  //gulp.start('grunt-build:dev', done);
```
Comment (crookedneighbor) [inline tasks/gulp-build.js]: This should be removed.

File: tasks/gulp-clean.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import gulp from 'gulp';
+
+let rimraf = require('rimraf');
+
+gulp.task('clean', (cb) => {
+	rimraf('website/build', cb);
```
Comment (crookedneighbor) [inline tasks/gulp-clean.js]: Our style is two spaces, not 4.

File: tasks/gulp-copy.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import gulp from 'gulp';
+
+gulp.task('copy', (cb) => {
+	gulp.src('website/public/favicon.ico').pipe(gulp.dest('website/build'));
```
Comment (crookedneighbor) [inline tasks/gulp-copy.js]: 2 spaces, not 4.

**Comment by crookedneighbor** [comment]
This is looking pretty good, just got some style notes for you.

Is there a reason you haven't deleted the other unused tasks for the gruntfile?

**Comment by noppanit** [comment]
Sorry I forgot to clean up, I'll do it and push.

**Comment by crookedneighbor** [comment]
What's the status on this?

**Comment by noppanit** [comment]
It's going to be pushed soon

**Comment by noppanit** [comment]
The build is failed because 

```
Error: Cannot find module 'xtend'
```

on module `through2` Have you seen this before?

**Comment by crookedneighbor** [comment]
No. Will look at it this weekend.

File: tasks/gulp-clean.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import gulp from 'gulp';
+
+let rimraf = require('rimraf');
```
Comment (crookedneighbor) [inline tasks/gulp-clean.js]: This should be `import rimraf from 'rimraf';`

**Comment by noppanit** [comment]
Related issue with Travis Failed on CoffeeScript https://github.com/github/generator-hubot/pull/6

File: tasks/gulp-build.js [hunk 6]
```diff
@@ -20,3 +54,5 @@ gulp.task('build:dev:watch', ['build:dev'], () => {
 gulp.task('build:prod', ['babel:common', 'prepare:staticNewStuff'], (done) => {
   gulp.start('grunt-build:prod', done);
 });
+
```
Comment (crookedneighbor) [inline tasks/gulp-build.js:27]: These two extra new lines are not necessary.

**Comment by noppanit** [comment]
I have two tasked that I have converted as I see that there's some test tasks and karma in https://github.com/HabitRPG/habitrpg/blob/develop/tasks/gulp-tests.js already. Not sure if the tasks in Gruntfile.js still being used.

File: tasks/gulp-build.js [hunk 7]
```diff
@@ -18,5 +34,95 @@ gulp.task('build:dev:watch', ['build:dev'], () => {
...
+  doBrowserify();
+  doUglify(uglifyFiles);
+  doCssmin(cssminFiles);
+  doStylus();
+  gulp.start('copy');
+  gulp.start('hashres');
 });
+
+
+function doBrowserify() {
```
Comment (crookedneighbor) [inline tasks/gulp-build.js]: Why did you make these normal functions instead of gulp tasks?

Comment (noppanit) [inline tasks/gulp-build.js]: At first I thought if would be easier, but thinking about it. It's probable easier to read if I make it a gulp task. I'll refactor it. And the style here is one file per gulp task?

Comment (crookedneighbor) [inline tasks/gulp-build.js]: For this case, yes, I think that'd be best.

**Comment by noppanit** [comment]
I've separated all the tasks to smaller ones. I'm not sure why the build is broken.

**Comment by noppanit** [comment]
@crookedneighbor you mind code review this PR?

File: tasks/gulp-uglify.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import gulp from 'gulp';
+import fs from 'fs';
+
+import {basename, dirname } from "./taskHelper.js";
+
+let rename = require("gulp-rename");
```
Comment (crookedneighbor) [inline tasks/gulp-uglify.js]: Please use the import syntax for all of these requires.

File: tasks/gulp-tests.js [hunk 9]
```diff
@@ -55,13 +55,21 @@ gulp.task('test:prepare:server', ['test:prepare:mongo'], () => {
 });
 
 gulp.task('test:prepare:build', (cb) => {
-  exec(testBin('grunt build:test'), cb);
+  exec(testBin('gulp test:prepare:translations && gulp build:dev'), cb);
```
Comment (crookedneighbor) [inline tasks/gulp-tests.js:68]: Why this instead of:

``` js
gulp.task('test:prepare:build', ['test:prepare:translations', 'build:dev']);
```

Comment (noppanit) [inline tasks/gulp-tests.js:68]: I tried that and somehow the build broke. I'll try that again.

**Comment by crookedneighbor** [comment]
Something is wrong with the css compiling:

screen shot 2015-10-30 at 1 12 44 pm

**Comment by crookedneighbor** [comment]
Any progress?

**Comment by noppanit** [comment]
Sorry was out of town. Should be done today or tmr. :-)

**Comment by noppanit** [comment]
@crookedneighbor apologies I'm pretty new at this project. I'm not sure what's causing the wrong css.  Can you help checking? I might be that I might have missed something.

**Comment by noppanit** [comment]
After I compare the new css, the positions for sprites are off

**Comment by gauthamchandra** [comment]
@noppanit, @crookedneighbor I am not too familiar with the codebase but I am with all the technologies and the stack involved. Tell me if you need help with this.

**Comment by gauthamchandra** [comment]
There needs to be a lot of cleanup with all the npm modules we use. Node is up to v5.0 and many of the modules we use requires `v0.10.x` and give a ton of deprecated warnings.

Once this PR is merged, I will file a new issue and a matching PR to clean up our dependencies and do a full check to make sure our code is properly linted and follows best practices

**Comment by paglias** [comment]
@gauthamchandra thanks for help!

A few things to keep in mind, though:
1. We're going to upgrade to Node v4.2, which is an LTS, not to Node 5. We might review it later but for the time being we'll switch to Node 4.2
2. We're in the process of upgrading most dependencies, switching to ES6 with ESLint for code style and re-writing the API completely in a separate PR https://github.com/HabitRPG/habitrpg/pull/6144
3. We'll be happy to accept any PR that upgrades dependencies to the newest version as long as it doesn't break compatibility with API v2 and hasn't already been addressed in https://github.com/HabitRPG/habitrpg/pull/6144

Let me know if you have any question

**Comment by noppanit** [comment]
@gauthamchandra I'm going to merge the conflict so please feel free to jump in this PR. I have two problems
1. https://github.com/HabitRPG/habitrpg/pull/6112/files#diff-565c7aaf8560ac8a12c341fa07bed43aR60 I can't seem to use 

`gulp.task('test:prepare:build', ['test:prepare:translations', 'build:dev']);` in Travis. The process wouldn't exit I suspect it's because some of the tasks hanging and wouldn't exit and Travis killed the build.
1. I can't seem to figure out the discrepancy of the `cssmin` between `gulp` and `grunt`. The generated css seems to be broken in `gulp`

Thanks for your help

**Comment by gauthamchandra** [comment]
@paglias wow that is a massive set of changes in the new PR. I will forgo the changes and look into that PR instead. :+1:

**Comment by noppanit** [comment]
@gauthamchandra Sorry I was on the road so didn't have a chance to catch up. Were you able to take a look at the PR?

**Comment by gauthamchandra** [comment]
@noppanit I got too busy over the weekend and wasn't able to look at it. I will try looking over it in the next 1-2 days.

**Comment by noppanit** [comment]
@crookedneighbor I think this PR is not going to work anymore as it's far behind the master. Is it ok if I create a new PR?

**Comment by crookedneighbor** [comment]
Let's do smaller PRs extracting one grunt task at a time.

**Comment by noppanit** [comment]
Will do that.

**Pull request closed without merging**